ambry-api/src/main/java/com.github.ambry/network/BoundedNettyByteBufReceive.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.network;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A netty {@link ByteBuf} version of Receive to buffer the incoming request or response.
+ */
+public class BoundedNettyByteBufReceive implements BoundedReceive<ByteBuf> {
+
+  private ByteBuf buffer = null;
+  private ByteBuf sizeBuffer = null;
+  private long sizeToRead = 0;
+  private long sizeRead = 0;
+  private Logger logger = LoggerFactory.getLogger(getClass());
+  private AtomicBoolean released = new AtomicBoolean(false);
+
+  public BoundedNettyByteBufReceive() {
+  }
+
+  @Override
+  public boolean isReadComplete() {
+    return !(buffer == null || sizeRead < sizeToRead);
+  }
+
+  /**
+   * Reading bytes from the {@link ReadableByteChannel} and writing those bytes to the given {@link ByteBuf}.
+   * This method would change the writerIndex of the given buffer.
+   * @param channel The {@link ReadableByteChannel} to read the bytes out.
+   * @param buffer The {@link ByteBuf} to write bytes to.
+   * @return The number of bytes read from channel.
+   * @throws IOException Any I/O error.
+   */
+  private int readBytesFromReadableByteChannel(ReadableByteChannel channel, ByteBuf buffer) throws IOException {
+    int n = channel.read(buffer.internalNioBuffer(buffer.writerIndex(), buffer.capacity() - buffer.writerIndex()));
```

Comment:
I suppose use nioBuffer is better than internalNioBuffer since it's for internal usage. And the reason this method is private and not exposed, is that I want to make sure all the ByteBuf it receives is not composite buffer and allocated where we want it to. We can make an public method to deal with more generic ByteBuf later.